### PR TITLE
Add `try_canonicalize` and use it over `std::fs::canonicalize`

### PR DESCRIPTION
**Pull request by Zoxc**
This adds a `try_canonicalize` function that calls `std::fs::canonicalize` and on Windows falls back to getting an absolute path. Uses of `canonicalize` have been replaced with `std::fs::canonicalize`. On Windows `std::fs::canonicalize` may fail due to incomplete drivers. In particular `ImDisk` does not support it.

Combined with https://github.com/rust-lang/rust/pull/109231 this allows compiling crates on an `ImDisk` RAM disk and I've tested that it works with various configuration using [rcb](https://github.com/Zoxc/rcb).

### CODE REVIEW

File: src/cargo/util/mod.rs [hunk 1]
```diff
@@ -133,6 +134,77 @@ pub fn truncate_with_ellipsis(s: &str, max_width: usize) -> String {
...
+pub fn try_canonicalize<P: AsRef<Path>>(path: P) -> std::io::Result<PathBuf> {
+    use std::ffi::OsString;
+    use std::io::Error;
+    use std::os::windows::ffi::{OsStrExt, OsStringExt};
+    use std::{io::ErrorKind, ptr};
+    use windows_sys::Win32::Foundation::{GetLastError, SetLastError};
+    use windows_sys::Win32::Storage::FileSystem::GetFullPathNameW;
+
+    // On Windows `canonicalize` may fail, so we fall back to getting an absolute path.
+    std::fs::canonicalize(&path).or_else(|_| {
```
Comment (weihanglo) [inline src/cargo/util/mod.rs:154]: Should we wait for the stabilization of `std::path::aboslute`? We should also have a TODO here and call out where this piece of code come from. (Looks like [`fn absolute`](https://github.com/rust-lang/rust/blob/da7c50c089d5db2d3ebaf227fe075bb1346bfaec/library/std/src/sys/windows/path.rs#L308-L333) on windows to me.)

**Review by weihanglo — COMMENTED** [review]
I am happy to approve this if Windows expert like @ChrisDenton nod on the Rust side. It will be better if std exposes a better API for this.

**Comment by epage** [comment]
> On Windows std::fs::canonicalize may fail due to incomplete drivers. In particular ImDisk does not support it.

Is there any issue discussion this?   This sounds like a lot larger of a pitfall and it seems like this should either be fixed in the stdlib or that there should be a general crate for solving this problem (similar to `dunce` or maybe a contribution to it)

File: src/cargo/util/mod.rs [hunk 2]
```diff
@@ -133,6 +134,77 @@ pub fn truncate_with_ellipsis(s: &str, max_width: usize) -> String {
...
+        unsafe {
+            SetLastError(0);
+            let len = GetFullPathNameW(path_u16.as_ptr(), 0, &mut [] as *mut u16, ptr::null_mut());
+            if len == 0 {
+                let error = GetLastError();
+                if error != 0 {
+                    return Err(Error::from_raw_os_error(error as i32));
+                }
+            }
+            let mut result: Vec<u16> = std::iter::repeat(0).take(len as usize).collect();
```
Comment (arlosi) [inline src/cargo/util/mod.rs:180]:
```suggestion
            let mut result = vec![0u16; len as usize];
```

File: src/cargo/util/mod.rs (hunk 1 above)
Comment (Zoxc) [reply src/cargo/util/mod.rs:154]: That would probably take a while.

**Comment by ehuss** [comment]
Thanks, this seems reasonable to me.  There is more long-term work to do here, but this should be fine to unblock the immediate needs. We have been using [`normalize_path`](https://github.com/rust-lang/cargo/blob/39684ff0aa88fe856460bc8665fd6f649b065e94/crates/cargo-util/src/paths.rs#L82) instead of `canonicalize` in many places, usually for this very same problem. We should probably eventually remove `normalize_path` and either use this or something else. However, `normalize_path` is used in many places, and it would be a pretty significant time commitment to understand the impact on all of them. 

I pushed a small spelling fix.

@bors r+

**Pull request merged**